File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.java
class Solution {
    public int numberOfSubarrays(int[] nums, int k) {

        int n = nums.length;

        Map<Integer,Integer> mp = new HashMap<>();

        int currSum = 0, res = 0;

        mp.put(currSum,1);

        for(int i = 0;i<n;i++){
            currSum += nums[i]%2;

            if(mp.containsKey(currSum-k)){
                res += mp.get(currSum-k);
            }

            mp.put(currSum,mp.getOrDefault(currSum,0) + 1);
        }

        return res;


        
    }
}

File: 1370-count-number-of-nice-subarrays/count-number-of-nice-subarrays.cpp
class Solution {
public:
    int numberOfSubarrays(vector<int>& nums, int k) {
        int n = nums.size();

        int cnt = 0;

        int i = 0;
        int j = 0;
        int ind = 0;

        int ans = 0;

        while(j<n){
            if(nums[j]%2){
                cnt++;
            }

            while(cnt>k){
                if(nums[i]%2){
                    cnt--;
                }
                i++;
                ind = i;
            }

            while(i<j && nums[i]%2==0){
                
                i++;
            }

            if(cnt==k){
                ans += (i - ind) + 1;
            }

            j++;
        }

        return ans;
    }
};